cosmwasm/cw20-token-minter/src/contract.rs

```diff
@@ -63,45 +72,57 @@ pub fn execute(
                 metadata,
                 subdenom: denom,
+                path,
+                channel,
+                token,
             } => {
-                // the first denom is always the same as the generated denom
-                DENOM_TO_BE_STORED.save(deps.storage, &denom)?;
-                let name = metadata.name;
-                // let symbol = metadata.symbol;
-                let msg = WasmMsg::Instantiate {
-                    admin: Some(env.contract.address.to_string()),
-                    code_id: config.cw20_base_code_id,
-                    label: denom.clone(),
-                    msg: to_json_binary(&cw20_base::msg::InstantiateMsg {
-                        // metadata is not guaranteed to always contain a name, however cw20_base::instantiate requires it to be set. if it is missing, we use the symbol instead.
-                        name: if name.is_empty() || name.len() > 50 {
-                            denom
-                        } else {
-                            name
+                Response::new()
+                    .add_message(
+                        // Instantiating the dummy contract first to be able to get the deterministic address
+                        WasmMsg::Instantiate2 {
+                            admin: Some(env.contract.address.to_string()),
+                            code_id: config.dummy_code_id,
+                            label: denom.clone(),
+                            msg: to_json_binary(&cosmwasm_std::Empty {})?,
+                            funds: vec![],
+                            salt: Binary::new(calculate_salt(
+                                U256::from_be_bytes::<{ U256::BYTES }>(
+                                    path.as_slice().try_into().expect("correctly encoded; qed"),
+                                ),
+                                channel,
+                                token.to_vec(),
+                            )),
                         },
-                        symbol: "ZKGM".to_string(),
-                        decimals: 0,
-                        initial_balances: vec![],
-                        mint: Some(cw20::MinterResponse {
-                            minter: env.contract.address.to_string(),
-                            cap: None,
-                        }),
-                        marketing: None,
-                    })?,
-                    funds: vec![],
-                };
-                Response::new().add_submessage(SubMsg::reply_on_success(msg, 1))
+                    )
+                    .add_message(
+                        // Then migrating to the actual `cw20_base` contract. Note that this contract has a custom
+                        // migrate entrypoint where it expects `InstantiateMsg` and calls the its `instantiate` function
+                        // in the `migrate` function
+                        WasmMsg::Migrate {
+                            contract_addr: denom.clone(),
```

Comment:
Maybe I missed something but isn't this supposed to be `instantiate2_address(...)` ?